url969d1ddca4_compmstr_java_scheme_tgz-pJ8-com_undi_javascheme_JavaSchemeJ8/src/com/undi/javascheme/SchemeReader.java

```diff
@@ -0,0 +1,427 @@
+package com.undi.javascheme;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PushbackInputStream;
+
+public class SchemeReader {
+  private PushbackInputStream mBufIn;
+  //Allow up to 10 characters to be pushed back into the buffer
+  private static final int PUSHBACK_SIZE = 10;
+  private ReadFunc[] readFuncs;
+  
+  private interface ReadFunc{
+    SchemeObject read();
+  }
+  
+  public static SchemeReader fromString(String in){
+    return new SchemeReader(new ByteArrayInputStream(in.getBytes()));
+  }
+  
+  public SchemeReader(InputStream in){
+    this.mBufIn = new PushbackInputStream(in, PUSHBACK_SIZE);
+    this.readFuncs = new ReadFunc[SchemeObject.type.NUM_TYPES.ordinal()];
+
+    this.readFuncs[SchemeObject.type.BOOLEAN.ordinal()] = new ReadFunc(){
+      public SchemeObject read(){ return readBoolean();} 
+    };
+    this.readFuncs[SchemeObject.type.NUMBER.ordinal()] = new ReadFunc(){
+      public SchemeObject read(){ return readNumber();} 
+    };
+    this.readFuncs[SchemeObject.type.CHARACTER.ordinal()] = new ReadFunc(){
+      public SchemeObject read(){ return readCharacter();} 
+    };
+    this.readFuncs[SchemeObject.type.STRING.ordinal()] = new ReadFunc(){
+      public SchemeObject read(){ return readString();} 
+    };
+    this.readFuncs[SchemeObject.type.SYMBOL.ordinal()] = new ReadFunc(){
+      public SchemeObject read(){ return readSymbol();} 
+    };
+    this.readFuncs[SchemeObject.type.PAIR.ordinal()] = new ReadFunc(){
+      public SchemeObject read(){ return readPair();} 
+    };
+    this.readFuncs[SchemeObject.type.EMPTY_LIST.ordinal()] = new ReadFunc(){
+      public SchemeObject read(){
+        //clean up trailing ')'
+        getc();
+        return SchemeObject.THE_EMPTY_LIST;
+        } 
+    };
+    this.readFuncs[SchemeObject.type.KEYWORD.ordinal()] = new ReadFunc() {
+		@Override
+		public SchemeObject read() { return readKeyword(); }
+	};
+    
+  }
+
+		/**
+		 * clears the rest of the input
+		 */
+		public void clearInput(){
+				try{
+						mBufIn.skip(mBufIn.available());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Check the number of bytes that were skipped when you called `skip()`. Handle cases when fewer than the expected number of bytes were skipped. See the following diffs on GitHub that make this code change: [diff1](https://github.com/apache/tomcat80/commit/facf983d8a7952e8d6b03204cb6610fc5f981dc6#diff-e58d7c084e48bd454de5beb71712f72d243bd0287ae2096d7b771f3a2dae6fb0R174), [diff2](https://github.com/apache/hadoop/commit/45620eee687b1093c812873a99833dab068e43d2#diff-f1ae2a4b3016a9deb983f2562d1e71372baf960cf5122d474a37a00f0307edb2R45)

Learn more: [InputStream.skip in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#skip-long-)